File: LocalPayment/src/main/java/com/braintreepayments/api/LocalPaymentRequest.java
package com.braintreepayments.api;

import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Builder used to construct an local payment request.
 */
public class LocalPaymentRequest {
    private static final String INTENT_KEY = "intent";
    private static final String RETURN_URL_KEY = "returnUrl";
    private static final String CANCEL_URL_KEY = "cancelUrl";
    private static final String EXPERIENCE_PROFILE_KEY = "experienceProfile";
    private static final String NO_SHIPPING_KEY = "noShipping";
    private static final String DISPLAY_NAME_KEY = "brandName";
    private static final String FUNDING_SOURCE_KEY = "fundingSource";
    private static final String AMOUNT_KEY = "amount";
    private static final String CURRENCY_CODE_KEY = "currencyIsoCode";
    private static final String GIVEN_NAME_KEY = "firstName";
    private static final String SURNAME_KEY = "lastName";
    private static final String EMAIL_KEY = "payerEmail";
    private static final String PHONE_KEY = "phone";
    private static final String STREET_ADDRESS_KEY = "line1";
    private static final String EXTENDED_ADDRESS_KEY = "line2";
    private static final String LOCALITY_KEY = "city";
    private static final String REGION_KEY = "state";
    private static final String POSTAL_CODE_KEY = "postalCode";
    private static final String COUNTRY_CODE_KEY = "countryCode";
    private static final String MERCHANT_ACCOUNT_ID_KEY = "merchantAccountId";
    private static final String PAYMENT_TYPE_COUNTRY_CODE_KEY = "paymentTypeCountryCode";
    private static final String BIC_KEY = "bic";

    private PostalAddress address;
    private String amount;
    private String bankIdentificationCode;
    private String currencyCode;
    private String displayName;
    private String email;
    private String givenName;
    private String merchantAccountId;
    private String paymentType;
    private String paymentTypeCountryCode;
    private String phone;
    private boolean shippingAddressRequired;
    private String surname;

    /**
     * @param address Optional - The address of the customer. An error will occur if this address is not valid.
     */
    public void setAddress(@Nullable PostalAddress address) {
        this.address = address;
    }

    /**
     * @param amount Optional - The amount for the transaction.
     */
    public void setAmount(@Nullable String amount) {
        this.amount = amount;
    }

    /**
     * @param bankIdentificationCode Optional - the Bank Identification Code of the customer (specific to iDEAL transactions).
     */
    public void setBic(@Nullable String bankIdentificationCode) {
        this.bankIdentificationCode = bankIdentificationCode;
    }

    /**
     * @param currencyCode Optional - A valid ISO currency code to use for the transaction. Defaults to merchant
     * currency code if not set.
     */
    public void setCurrencyCode(@Nullable String currencyCode) {
        this.currencyCode = currencyCode;
    }

    /**
     * @param displayName Optional - The merchant name displayed inside of the Local Payment flow
     *                    when starting the payment.
     */
    public void setDisplayName(@Nullable String displayName) { this.displayName = displayName; }

    /**
     * @param email Optional - Payer email of the customer.
     */
    public void setEmail(@Nullable String email) {
        this.email = email;
    }

    /**
     * @param givenName Optional - Given (first) name of the customer.
     */
    public void setGivenName(@Nullable String givenName) {
        this.givenName = givenName;
    }

    /**
     * @param merchantAccountId Optional - A non-default merchant account to use for tokenization.
     */
    public void setMerchantAccountId(@Nullable String merchantAccountId) {
        this.merchantAccountId = merchantAccountId;
    }

    /**
     * @param paymentType - The type of payment
     */
    public void setPaymentType(@Nullable String paymentType) {
        this.paymentType = paymentType;
    }

    /**
     * @param paymentTypeCountryCode The country code of the local payment. This value must be one of
     *                               the supported country codes for a given local payment type listed.
     *                               For local payments supported in multiple countries, this value
     *                               may determine which banks are presented to the customer.
     *                               @see <a href=https://developers.braintreepayments.com/guides/local-payment-methods/client-side-custom/android/v3#invoke-payment-flow>Supported Country Codes</a>
     */
    public void setPaymentTypeCountryCode(@Nullable String paymentTypeCountryCode) {
        this.paymentTypeCountryCode = paymentTypeCountryCode;
    }

    /**
     * @param phone Optional - Phone number of the customer.
     */
    public void setPhone(@Nullable String phone) {
        this.phone = phone;
    }

    /**
     * @param shippingAddressRequired - Indicates whether or not the payment needs to be shipped. For digital goods,
     *                                this should be false. Defaults to false.
     */
    public void setShippingAddressRequired(boolean shippingAddressRequired) {
        this.shippingAddressRequired = shippingAddressRequired;
    }

    /**
     * @param surname Optional - Surname (last name) of the customer.
     */
    public void setSurname(@Nullable String surname) {
        this.surname = surname;
    }

    @Nullable
    public PostalAddress getAddress() {
        return address;
    }

    @Nullable
    public String getAmount() {
        return amount;
    }

    @Nullable
    public String getBic() {
        return bankIdentificationCode;
    }

    @Nullable
    public String getCurrencyCode() {
        return currencyCode;
    }

    @Nullable
    public String getDisplayName() {
        return displayName;
    }

    @Nullable
    public String getEmail() {
        return email;
    }

    @Nullable
    public String getGivenName() {
        return givenName;
    }

    @Nullable
    public String getMerchantAccountId() {
        return merchantAccountId;
    }

    @Nullable
    public String getPaymentType() {
        return paymentType;
    }

    @Nullable
    public String getPaymentTypeCountryCode() {
        return paymentTypeCountryCode;
    }

    @Nullable
    public String getPhone() {
        return phone;
    }

    public boolean isShippingAddressRequired() {
        return shippingAddressRequired;
    }

    @Nullable
    public String getSurname() {
        return surname;
    }

    public String build(String returnUrl, String cancelUrl) {
        try {
            JSONObject payload = new JSONObject()
                    .put(INTENT_KEY, "sale")
                    .put(RETURN_URL_KEY, returnUrl)
                    .put(CANCEL_URL_KEY, cancelUrl)
                    .put(FUNDING_SOURCE_KEY, paymentType)
                    .put(AMOUNT_KEY, amount)
                    .put(CURRENCY_CODE_KEY, currencyCode)
                    .put(GIVEN_NAME_KEY, givenName)
                    .put(SURNAME_KEY, surname)
                    .put(EMAIL_KEY, email)
                    .put(PHONE_KEY, phone)
                    .put(MERCHANT_ACCOUNT_ID_KEY, merchantAccountId)
                    .putOpt(PAYMENT_TYPE_COUNTRY_CODE_KEY, paymentTypeCountryCode)
                    .putOpt(BIC_KEY, bankIdentificationCode);

            if (address != null) {
                payload.put(STREET_ADDRESS_KEY, address.getStreetAddress())
                        .put(EXTENDED_ADDRESS_KEY, address.getExtendedAddress())
                        .put(LOCALITY_KEY, address.getLocality())
                        .put(REGION_KEY, address.getRegion())
                        .put(POSTAL_CODE_KEY, address.getPostalCode())
                        .put(COUNTRY_CODE_KEY, address.getCountryCodeAlpha2());
            }

            JSONObject experienceProfile = new JSONObject();
            experienceProfile.put(NO_SHIPPING_KEY, !shippingAddressRequired);
            experienceProfile.put(DISPLAY_NAME_KEY, displayName);
            payload.put(EXPERIENCE_PROFILE_KEY, experienceProfile);

            return payload.toString();
        } catch (JSONException ignored) {}

        return new JSONObject().toString();
    }
}

File: LocalPayment/src/main/java/com/braintreepayments/api/localpayment/LocalPaymentRequest.kt
package com.braintreepayments.api.localpayment

import com.braintreepayments.api.core.PostalAddress
import org.json.JSONException
import org.json.JSONObject

/**
 * Builder used to construct an local payment request.
 *
 * @param hasUserLocationConsent is an optional parameter that informs the SDK
 *  if your application has obtained consent from the user to collect location data in compliance with
 *  [Google Play Developer Program policies](https://support.google.com/googleplay/android-developer/answer/10144311#personal-sensitive)
 *  This flag enables PayPal to collect necessary information required for Fraud Detection and Risk Management.
 *
 *  @see [User Data policies for the Google Play Developer Program ](https://support.google.com/googleplay/android-developer/answer/10144311.personal-sensitive)
 *
 *  @see [Examples of prominent in-app disclosures](https://support.google.com/googleplay/android-developer/answer/9799150?hl=en.Prominent%20in-app%20disclosure)
 * @property address Optional - The address of the customer. An error will occur if this address
 *                              is not valid.
 * @property amount Optional - The amount for the transaction.
 * @property bankIdentificationCode Optional - the Bank Identification Code
 *                                  of the customer (specific to iDEAL transactions).
 * @property currencyCode Optional - A valid ISO currency code to use for the transaction.
 *                        Defaults to merchant currency code if not set.
 * @property displayName Optional - The merchant name displayed inside of the Local Payment flow
 *                       when starting the payment.
 * @property email Optional - Payer email of the customer.
 * @property givenName Optional - Given (first) name of the customer.
 * @property merchantAccountId Optional - A non-default merchant account to use for tokenization.
 * @property paymentType - The type of payment
 * @property paymentTypeCountryCode The country code of the local payment. This value must be one of
 *       the supported country codes for a given local payment type listed.
 *       For local payments supported in multiple countries, this value
 *       may determine which banks are presented to the customer.
 *       @see [Supported Country Codes](https://developer.paypal.com/braintree/docs/guides/local-payment-methods/client-side-custom/android/v4.invoke-payment-flow)
 * @property phone Optional - Phone number of the customer.
 * @property isShippingAddressRequired - Indicates whether or not the payment needs to be shipped.
 *                                    For digital goods, this should be false. Defaults to false.
 * @property surname Optional - Surname (last name) of the customer.
 */
data class LocalPaymentRequest @JvmOverloads constructor(
    val hasUserLocationConsent: Boolean,
    var address: PostalAddress? = null,
    var amount: String? = null,
    var bankIdentificationCode: String? = null,
    var currencyCode: String? = null,
    var displayName: String? = null,
    var email: String? = null,
    var givenName: String? = null,
    var merchantAccountId: String? = null,
    var paymentType: String? = null,
    var paymentTypeCountryCode: String? = null,
    var phone: String? = null,
    var isShippingAddressRequired: Boolean = false,
    var surname: String? = null
) {

    fun build(returnUrl: String?, cancelUrl: String?): String {
        try {
            val payload = JSONObject()
                .put(INTENT_KEY, "sale")
                .put(RETURN_URL_KEY, returnUrl)
                .put(CANCEL_URL_KEY, cancelUrl)
                .put(FUNDING_SOURCE_KEY, paymentType)
                .put(AMOUNT_KEY, amount)
                .put(CURRENCY_CODE_KEY, currencyCode)
                .put(GIVEN_NAME_KEY, givenName)
                .put(SURNAME_KEY, surname)
                .put(EMAIL_KEY, email)
                .put(PHONE_KEY, phone)
                .put(MERCHANT_ACCOUNT_ID_KEY, merchantAccountId)
                .putOpt(PAYMENT_TYPE_COUNTRY_CODE_KEY, paymentTypeCountryCode)
                .putOpt(BIC_KEY, bankIdentificationCode)

            if (address != null) {
                payload.put(STREET_ADDRESS_KEY, address!!.streetAddress)
                    .put(EXTENDED_ADDRESS_KEY, address!!.extendedAddress)
                    .put(LOCALITY_KEY, address!!.locality)
                    .put(REGION_KEY, address!!.region)
                    .put(POSTAL_CODE_KEY, address!!.postalCode)
                    .put(COUNTRY_CODE_KEY, address!!.countryCodeAlpha2)
            }

            val experienceProfile = JSONObject()
            experienceProfile.put(NO_SHIPPING_KEY, !isShippingAddressRequired)
            experienceProfile.put(DISPLAY_NAME_KEY, displayName)
            payload.put(EXPERIENCE_PROFILE_KEY, experienceProfile)

            return payload.toString()
        } catch (ignored: JSONException) {
        }

        return JSONObject().toString()
    }

    companion object {
        private const val INTENT_KEY = "intent"
        private const val RETURN_URL_KEY = "returnUrl"
        private const val CANCEL_URL_KEY = "cancelUrl"
        private const val EXPERIENCE_PROFILE_KEY = "experienceProfile"
        private const val NO_SHIPPING_KEY = "noShipping"
        private const val DISPLAY_NAME_KEY = "brandName"
        private const val FUNDING_SOURCE_KEY = "fundingSource"
        private const val AMOUNT_KEY = "amount"
        private const val CURRENCY_CODE_KEY = "currencyIsoCode"
        private const val GIVEN_NAME_KEY = "firstName"
        private const val SURNAME_KEY = "lastName"
        private const val EMAIL_KEY = "payerEmail"
        private const val PHONE_KEY = "phone"
        private const val STREET_ADDRESS_KEY = "line1"
        private const val EXTENDED_ADDRESS_KEY = "line2"
        private const val LOCALITY_KEY = "city"
        private const val REGION_KEY = "state"
        private const val POSTAL_CODE_KEY = "postalCode"
        private const val COUNTRY_CODE_KEY = "countryCode"
        private const val MERCHANT_ACCOUNT_ID_KEY = "merchantAccountId"
        private const val PAYMENT_TYPE_COUNTRY_CODE_KEY = "paymentTypeCountryCode"
        private const val BIC_KEY = "bic"
    }
}
